### ISSUE TITLE
Java APM Agent 1.34.1 refactor broke span tracking in a Scala app

### ISSUE THREAD

**Issue by bplies-ATX** [elastic/apm-agent-java#3052]
## Describe the bug
When we upgraded from `1.33.0` to `1.36.0` in our Scala application we noticed that tracking of spans stopped tracking correctly and could be easily seen in the loss of diversity in "span types" as well as a flattening of traces. We spotted a [change](https://github.com/elastic/apm-agent-java/pull/2783/files) in `1.34.1` that gives us some concern. We downgraded to `1.34.0` (still an upgrade from `1.33.0`) and the problem went away. We upgraded to `1.34.1` and the problem appeared again.  

Strangely the effect is not immediate and can take a few minutes to hours of runtime in order to manifest.  When applied to a lower testing environment with much less traffic volume, we cannot reproduce the span tracking problem.

For now we're locked at version `1.34.0` with no ability to upgrade until this is fixed.  Although Scala isn't officially supported by Java APM, these recent versions are, in our case, becoming _less_ compatible rather than more.  There is an amount of manual transaction/span creation occurring via the APM API in order to smooth over the differences in threading model. 

See Elastic Support case #01335141 for screenshots. 

Only other possible clues we have are some observed errors from the APM Server complaining about issues with transactions that it was receiving from the APM agent.

## Steps to reproduce

- Update the Agent version to `1.34.1+` 
- Deploy to a well-trafficked environment
- Wait some amount of time
- Spans stop being reported

## Expected behavior
Bumping to a newer APM Agent version shouldn't substantially break manual instrumentations.

## Debug logs
We are not able to provide this.

**Comment by bplies-ATX** [comment]
One clarification:  We upgrade both the APM agent attached to the JVM process _as well as_ the library dependency for `com.elastic.apm:apm-agent-api:${version}` so that they both agree.

**Comment by jackshirazi** [comment]
The diff is https://github.com/elastic/apm-agent-java/compare/v1.34.0...v1.34.1

**Comment by jackshirazi** [comment]
You could try testing with a higher value of [transaction_max_spans](https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-transaction-max-spans) in case it is related to the activation stack limit?

**Comment by bplies-ATX** [comment]
> 
Thank you for the suggestion.  However the `transaction_max_spans` has a default of `500` and our transactions are much lower than that already.  In a shared screenshot the demonstrated previously working transaction was just 4 spans.

**Comment by jackshirazi** [comment]
do you see the message
```
... Activation stack depth reached its maximum ...
```
in the logs?

**Comment by bplies-ATX** [comment]
No. There is nothing about "activ*" on _any_ `log.level`

**Comment by JonasKunz** [comment]
Hi @bplies-ATX , the log is supposed to appear in the [agent logs](https://www.elastic.co/guide/en/apm/agent/java/current/trouble-shooting.html#trouble-shooting-logging-procedure) and not within the apm-server logs. Did you check the agent logs?

**Comment by bplies-ATX** [comment]
I am seeing several log statements of this pattern with APM Java Agent `1.34.0` where we still do get a lot of spans
```
2023-03-27 15:17:18,441 [dda-system-akka.actor.default-dispatcher-15] WARN  co.elastic.apm.agent.impl.ElasticApmTracer - Deactivating a context ('GET /ENDPOINT1' 00-5db974a9cc35ea0dfeea10a51dd9f4b1-986113757de0e9d5-01 (3da032d0)) which is not the currently active one ('GET /ENDPOINT2' 00-98db85ee12d9e0e4096c371fc91bcf81-2d15825eaf6b6ada-01 (b3d2ad5)). This can happen when not properly deactivating a previous span or context.
```

We weren't aware that there was a current problem. It has been exceedingly difficult to get APM to cooperate with the threading model present in the Scala stack.   We cannot go beyond this version or it'll rapidly get into a state where we won't get any spans at all.  We will try to fix what we can on our end to resolve the above error.

**Comment by JonasKunz** [comment]
Okay, in combination with what you said in your initial message this makes sense now:

> There is an amount of manual transaction/span creation occurring via the APM API in order to smooth over the differences in threading model

What I assume is happening is that your custom instrumentation happens to not properly deactivate spans when they are not executing anymore on a thread. To be more precise, I would assume that the number of activations is greater than the number of deactivations. This makes the stack of active contexts grow, which was limited to a fixed size in `1.34.1` to prevent memory leaks.

I would suggest you to use debug logs to find out where and why you are missing deactivations. Trace level logs can also help, because then stacktraces of the activations (due to reference increment / decrement) will be printed.

I'm closing this for now, as it seems to not be a bug within the agent but your custom instrumentation. Feel free to reopen in case I'm getting this wrong.
